commons/src/scanner/symbol/iterator/mod.rs

```diff
@@ -0,0 +1,504 @@
+        if index >= self.curr_index() {
+            match self.kind.borrow_mut() {
+                SymbolIteratorKind::Nested(parent) => parent.set_peek_index(index),
+                SymbolIteratorKind::Root(root) => {
+                    root.peek_index = index;
+                }
+            }
+        }
+    }
+
+    /// Resets peek to get `peek() == next()`.
+    ///
+    /// **Note:** Needed to reset peek index after using `peeking_next()`.
+    pub fn reset_peek(&mut self) {
+        self.set_peek_index(self.curr_index());
+    }
+
+    /// Returns the maximal remaining symbols in this iterator.
+    ///
+    /// **Note:** This slice does not consider parent iterators, or matching functions.
+    /// Therefore, the returned [`Symbol`] slice might differ from the symbols returned by calling [`Self::next()`],
+    /// but [`Self::next()`] cannot return more symbols than those inside the returned slice.
+    pub fn max_remaining_symbols(&self) -> Option<&'input [Symbol<'input>]> {
+        match &self.kind {
+            SymbolIteratorKind::Nested(parent) => parent.max_remaining_symbols(),
+            SymbolIteratorKind::Root(root) => root.remaining_symbols(),
+        }
+    }
+
+    /// Returns the next [`Symbol`] without changing the current index.    
+    pub fn peek(&mut self) -> Option<&'input Symbol<'input>> {
+        let symbol = self.peeking_next(|_| true);
+        self.reset_peek(); // Note: Resetting index, because peek() must be idempotent
+        symbol
+    }
+
+    /// Returns the [`SymbolKind`] of the peeked [`Symbol`].
+    pub fn peek_kind(&mut self) -> Option<SymbolKind> {
+        self.peek().map(|s| s.kind)
+    }
+
+    /// Nests this iterator, by creating a new iterator that has this iterator set as parent.
+    ///
+    /// **Note:** Any change in this iterator is **not** propagated to the nested iterator.
+    /// See [`Self::update()`] on how to synchronize this iterator with the nested one.
+    ///
+    /// # Arguments
+    ///
+    /// * `prefix_match` ... Optional matching function used to strip prefix on new lines
+    /// * `end_match` ... Optional matching function used to indicate the end of the created iterator
+    pub fn nest(
+        &self,
+        prefix_match: Option<IteratorPrefixFn>,
+        end_match: Option<IteratorEndFn>,
+    ) -> SymbolIterator<'input> {
+        SymbolIterator {
+            kind: SymbolIteratorKind::Nested(Box::new(self.clone())),
+            start_index: self.curr_index(),
+            depth: self.depth + 1,
+            prefix_match,
+            end_match,
+            iter_end: self.iter_end,
+        }
+    }
+
+    /// Updates the given parent iterator to take the progress of the nested iterator.
+    ///
+    /// **Note:** Only updates the parent if `self` is nested.
+    pub fn update(self, parent: &mut Self) {
+        if let SymbolIteratorKind::Nested(self_parent) = self.kind {
+            *parent = *self_parent;
+        }
```

Comment:
added comparison for `start_index`, because this is the closest and most cheapest equality check we can get. Because we clone the iterator on `nest()`, other indices will most likely differ.